### PR TITLE
Preemption for system jobs

### PR DESCRIPTION
**Pull request by preetapan**
This PR brings in prior work from PR #4710, rebased to master and modifying all the internal comparison logic to use the new `ComparableResources` struct.

### CODE REVIEW

File: scheduler/preemption.go [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+		distance2 := basicResourceDistance(resourcesNeeded, bestAllocs[j].ComparableResources())
+		return distance1 > distance2
+	})
+
+	filteredBestAllocs := eliminateSuperSetAllocationsForTaskGroup(bestAllocs, nodeRemainingResources, resourcesNeeded)
+	return filteredBestAllocs
+
+}
+
+// computeCurrentPreemptions counts the number of other allocations being preempted that match the job and task group of
```
Comment (preetapan) [inline scheduler/preemption.go]: todo this could be calculated ahead of time per jobid in the set of existing allocations on the node

File: scheduler/preemption.go [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+
+	//TODO(preetha): should add some debug logging
+
+	nodeRemainingResources := node.ComparableResources()
+
+	// Initialize nodeRemainingResources with the remaining resources
+	// after accounting for reserved resources and all allocations
+
+	// Subtract the reserved resources of the node
+	if node.ComparableReservedResources() != nil {
```
Comment (dadgar) [inline scheduler/preemption.go]: `if c := node.ComparableReservedResources(); c != nil`

Comment (preetapan) [reply scheduler/preemption.go]:
```suggestion
	if node.ComparableReservedResources() != nil {
```

File: nomad/operator_endpoint.go [hunk 3]
```diff
@@ -284,3 +284,64 @@ func (op *Operator) ServerHealth(args *structs.GenericRequest, reply *autopilot.
...
+
+	// Check if the return type is a bool.
+	if respBool, ok := resp.(bool); ok {
+		*reply = respBool
+	}
+	return nil
+}
+
+// SchedulerGetConfiguration is used to retrieve the current Scheduler configuration.
+func (op *Operator) SchedulerGetConfiguration(args *structs.GenericRequest, reply *structs.SchedulerConfiguration) error {
```
Comment (dadgar) [inline nomad/operator_endpoint.go]: Change to a response type as well

File: nomad/operator_endpoint.go [hunk 4]
```diff
@@ -284,3 +284,64 @@ func (op *Operator) ServerHealth(args *structs.GenericRequest, reply *autopilot.
...
+	_, config, err := state.SchedulerConfig()
+	if err != nil {
+		return err
+	}
+	if config == nil {
+		return fmt.Errorf("scheduler config not initialized yet")
+	}
+
+	*reply = *config
+
```
Comment (dadgar) [inline nomad/operator_endpoint.go]: Set indexes on the responses

File: nomad/state/schema.go [hunk 5]
```diff
@@ -599,3 +600,21 @@ func aclTokenTableSchema() *memdb.TableSchema {
...
+func schedulerConfigTableSchema() *memdb.TableSchema {
+	return &memdb.TableSchema{
+		Name: "scheduler_config",
+		Indexes: map[string]*memdb.IndexSchema{
+			"id": {
+				Name:         "id",
+				AllowMissing: true,
+				Unique:       true,
+				Indexer: &memdb.ConditionalIndex{
+					Conditional: func(obj interface{}) (bool, error) { return true, nil },
```
Comment (dadgar) [inline nomad/state/schema.go]: Can you put a comment since the reason for this isn't super obvious

File: nomad/state/state_store.go [hunk 6]
```diff
@@ -3956,3 +3991,81 @@ func (r *StateRestore) addEphemeralDiskToTaskGroups(job *structs.Job) {
 		}
 	}
 }
+
```
Comment (dadgar) [inline nomad/state/state_store.go]: Move the code above the restore functions

File: nomad/fsm.go [hunk 7]
```diff
@@ -1833,6 +1835,23 @@ func (s *nomadSnapshot) persistACLTokens(sink raft.SnapshotSink,
...
 
+func (n *nomadFSM) applySchedulerConfigUpdate(buf []byte, index uint64) interface{} {
+	var req structs.SchedulerSetConfigRequest
+	if err := structs.Decode(buf, &req); err != nil {
+		panic(fmt.Errorf("failed to decode request: %v", err))
+	}
+	defer metrics.MeasureSince([]string{"nomad", "fsm", "scheduler-config"}, time.Now())
+
+	if req.CAS {
+		act, err := n.state.SchedulerCASConfig(index, req.Config.ModifyIndex, &req.Config)
```
Comment (dadgar) [inline nomad/fsm.go]: act is a weird variable name? `applied`?

File: nomad/structs/operator.go [hunk 8]
```diff
@@ -119,3 +119,32 @@ type AutopilotConfig struct {
 	CreateIndex uint64
 	ModifyIndex uint64
 }
+
+type SchedulerConfiguration struct {
```
Comment (dadgar) [inline nomad/structs/operator.go]: Comment the struct

File: nomad/fsm.go [hunk 9]
```diff
@@ -1833,6 +1835,23 @@ func (s *nomadSnapshot) persistACLTokens(sink raft.SnapshotSink,
 	return nil
 }
 
+func (n *nomadFSM) applySchedulerConfigUpdate(buf []byte, index uint64) interface{} {
```
Comment (dadgar) [inline nomad/fsm.go]: Move to be above here: https://github.com/hashicorp/nomad/blob/4db7e87a31b842ace0c12081aa18229da376cffc/nomad/fsm.go#L998

Comment (dadgar) [inline nomad/fsm.go]: Where is the snapshot and restore code?

File: scheduler/system_sched.go [hunk 10]
```diff
@@ -347,6 +347,24 @@ func (s *SystemScheduler) computePlacements(place []allocTuple) error {
 				alloc.PreviousAllocation = missing.Alloc.ID
 			}
 
+			// If this placement involves preemption, set DesiredState to stop for those allocations
```
Comment (dadgar) [inline scheduler/system_sched.go]: stop -> evict

File: scheduler/rank.go [hunk 11]
```diff
@@ -211,10 +225,40 @@ OUTER:
...
+					// If eviction is not enabled, mark this node as exhausted and continue
+					if !iter.evict {
+						iter.ctx.Metrics().ExhaustedNode(option.Node,
+							fmt.Sprintf("network: %s", err))
+						netIdx.Release()
+						continue OUTER
+					}
+
+					// Look for preemptible allocations to satisfy the network resource for this task
+					preemptedAllocsForTaskNetwork := preemptForNetworkResourceAsk(iter.priority, proposed, ask, netIdx, currentPreemptions)
```
Comment (dadgar) [inline scheduler/rank.go]: Variable name is super long. `netPreemptions`

File: scheduler/rank.go [hunk 12]
```diff
@@ -211,10 +225,40 @@ OUTER:
...
 					netIdx.Release()
-					continue OUTER
+					netIdx = structs.NewNetworkIndex()
+					netIdx.SetNode(option.Node)
+					netIdx.AddAllocs(proposed)
+
+					offer, err = netIdx.AssignNetwork(ask)
+					if offer == nil {
+						iter.ctx.Metrics().ExhaustedNode(option.Node,
+							fmt.Sprintf("unexecpted error, unable to create offer after preempting:%v", err))
```
Comment (dadgar) [inline scheduler/rank.go]: Hm seems odd to expose this in the metrics. I would drop that an instead emit an error level log (`iter.ctx.Logger().Error(...)`)

File: scheduler/rank.go [hunk 13]
```diff
@@ -195,6 +199,16 @@ OUTER:
 				DiskMB: int64(iter.taskGroup.EphemeralDisk.SizeMB),
 			},
 		}
+
```
Comment (dadgar) [inline scheduler/rank.go]: I think this method has to be refactored. It will get even more complex once devices are considered. I had a hard to following the code and I know it quite well

File: scheduler/rank.go [hunk 14]
```diff
@@ -234,19 +278,32 @@ OUTER:
...
+				continue
+			}
+			// If eviction is enabled and the node doesn't fit the alloc, check if
+			// any allocs can be preempted
+
+			// Remove the last element containing the current placement from proposed allocs
+			current := proposed[:len(proposed)-1]
+			preemptForTaskGroup := findPreemptibleAllocationsForTaskGroup(iter.priority, current, total, option.Node, currentPreemptions)
+			allocsToPreempt = append(allocsToPreempt, preemptForTaskGroup...)
+			// If we were unable to find preempted allocs to meet these requirements
```
Comment (dadgar) [inline scheduler/rank.go]: Add whitespace between code and comments; line 290 as well

File: scheduler/rank.go [hunk 15]
```diff
@@ -234,19 +278,32 @@ OUTER:
...
+			// Skip the node if evictions are not enabled
+			if !iter.evict {
+				iter.ctx.Metrics().ExhaustedNode(option.Node, dim)
+				continue
+			}
+			// If eviction is enabled and the node doesn't fit the alloc, check if
+			// any allocs can be preempted
+
+			// Remove the last element containing the current placement from proposed allocs
+			current := proposed[:len(proposed)-1]
```
Comment (dadgar) [inline scheduler/rank.go]: Not a fan of this. Capture it outside as a variable

File: scheduler/system_sched.go [hunk 16]
```diff
@@ -347,6 +347,24 @@ func (s *SystemScheduler) computePlacements(place []allocTuple) error {
...
+			if option.PreemptedAllocs != nil {
```
Comment (dadgar) [inline scheduler/system_sched.go]: Would it be cleaner to call a method, called handleAllocWithPreemptions

File: nomad/operator_endpoint.go [hunk 17]
```diff
@@ -284,3 +284,64 @@ func (op *Operator) ServerHealth(args *structs.GenericRequest, reply *autopilot.
 
 	return nil
 }
+
+// SchedulerSetConfiguration is used to set the current Scheduler configuration.
+func (op *Operator) SchedulerSetConfiguration(args *structs.SchedulerSetConfigRequest, reply *bool) error {
```
[the remark this answers is not in the document]
Comment (preetapan) [reply nomad/operator_endpoint.go]: Why does the autopilotset endpoint do the same thing? I figured bool was fine because its basically only ever going to be true/false for the compare and set case.

File: nomad/operator_endpoint.go [hunk 18]
```diff
@@ -284,3 +284,64 @@ func (op *Operator) ServerHealth(args *structs.GenericRequest, reply *autopilot.
...
+	if done, err := op.srv.forward("Operator.SchedulerSetConfiguration", args, args, reply); done {
+		return err
+	}
+
+	// This action requires operator write access.
+	rule, err := op.srv.ResolveToken(args.AuthToken)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (preetapan) [reply nomad/operator_endpoint.go]: its not consistent in the rest of this file in other methods, but I don't mind changing this one for readability

File: nomad/state/state_store.go [hunk 19]
```diff
@@ -218,6 +218,41 @@ func (s *StateStore) UpsertPlanResults(index uint64, results *structs.ApplyPlanR
...
+		exist := existing.(*structs.Allocation)
+
+		// Copy everything from the existing allocation
+		copyAlloc := exist.Copy()
+
+		// Only update the fields set by the scheduler
+		copyAlloc.DesiredStatus = preemptedAlloc.DesiredStatus
+		copyAlloc.PreemptedByAllocation = preemptedAlloc.PreemptedByAllocation
+		copyAlloc.DesiredDescription = preemptedAlloc.DesiredDescription
+
```
[the remark this answers is not in the document]
Comment (preetapan) [reply nomad/state/state_store.go]: good catch

File: nomad/structs/structs.go [hunk 20]
```diff
@@ -7377,6 +7416,37 @@ func (a *Allocation) ComparableResources() *ComparableResources {
 	}
 }
 
+// COMPAT(0.11): Remove in 0.11
+// CompatibleNetworkResources returns network resources on the allocation
+// by reading AllocatedResources which are populated starting in 0.9 and
+// falling back to pre 0.9 fields (Resources/TaskResources) if set
+func (a *Allocation) CompatibleNetworkResources() []*NetworkResource {
+	var ret []*NetworkResource
```
[the remark this answers is not in the document]
Comment (preetapan) [reply nomad/structs/structs.go]: because ComparableResources ignores networks entirely. I didn't know how safe it was to change that given how much its used everywhere so made something very specific for preemption.

File: nomad/structs/structs.go [hunk 21]
```diff
@@ -7968,11 +8039,12 @@ func (e *Evaluation) ShouldBlock() bool {
...
-		Priority:       e.Priority,
-		Job:            j,
-		NodeUpdate:     make(map[string][]*Allocation),
-		NodeAllocation: make(map[string][]*Allocation),
+		EvalID:          e.ID,
+		Priority:        e.Priority,
+		Job:             j,
+		NodeUpdate:      make(map[string][]*Allocation),
+		NodeAllocation:  make(map[string][]*Allocation),
+		NodePreemptions: make(map[string][]*Allocation),
```
[the remark this answers is not in the document]
Comment (preetapan) [reply nomad/structs/structs.go]: Whats the downside to having an empty map, its easier than having to check for the map being nil everywhere

File: nomad/fsm_test.go [hunk 22]
```diff
@@ -2825,3 +2825,49 @@ func TestFSM_Autopilot(t *testing.T) {
 		t.Fatalf("bad: %v", config.CleanupDeadServers)
 	}
 }
+
```
[the remark this answers is not in the document]
Comment (preetapan) [reply nomad/fsm_test.go]: Done in line 1754 by modifying existing test

File: command/agent/operator_endpoint.go [hunk 23]
```diff
@@ -208,3 +208,73 @@ func (s *HTTPServer) OperatorServerHealth(resp http.ResponseWriter, req *http.Re
...
+
+		resp := api.SchedulerConfigurationResponse{
+			SchedulerConfig: out,
+			CreateIndex:     out.CreateIndex,
+			ModifyIndex:     out.ModifyIndex,
+		}
+
+		return resp, nil
+
+	case "PUT":
```
Comment (dadgar) [inline command/agent/operator_endpoint.go]: We accept POST and PUT

File: command/agent/operator_endpoint.go [hunk 24]
```diff
@@ -208,3 +208,73 @@ func (s *HTTPServer) OperatorServerHealth(resp http.ResponseWriter, req *http.Re
...
 	return out, nil
 }
+
+// OperatorSchedulerConfiguration is used to inspect the current Scheduler configuration.
+// This supports the stale query mode in case the cluster doesn't have a leader.
+func (s *HTTPServer) OperatorSchedulerConfiguration(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	// Switch on the method
+	switch req.Method {
+	case "GET":
+		var args structs.GenericRequest
```
Comment (dadgar) [inline command/agent/operator_endpoint.go]: Generally split the methods up so that it only handles one type of HTTP verb: https://github.com/hashicorp/nomad/blob/master/command/agent/job_endpoint.go#L15

File: api/operator.go [hunk 25]
```diff
@@ -106,3 +108,61 @@ func (op *Operator) RaftRemovePeerByID(id string, q *WriteOptions) error {
...
+	ModifyIndex uint64
+}
+
+// SchedulerConfigurationResponse is the response object that wraps SchedulerConfiguration
+type SchedulerConfigurationResponse struct {
+	// SchedulerConfig contains scheduler config options
+	SchedulerConfig SchedulerConfiguration
+
+	// CreateIndex/ModifyIndex store the create/modify indexes of this configuration.
+	CreateIndex uint64
```
Comment (dadgar) [inline api/operator.go]: Should be WriteMeta and QueryMeta: https://github.com/hashicorp/nomad/blob/master/api/jobs.go#L1039-L1046

File: command/agent/operator_endpoint.go [hunk 26]
```diff
@@ -208,3 +208,73 @@ func (s *HTTPServer) OperatorServerHealth(resp http.ResponseWriter, req *http.Re
...
+		}
+
+		// Only use the out value if this was a CAS
+		if !args.CAS {
+			return true, nil
+		}
+		return reply, nil
+
+	default:
+		return nil, CodedError(404, ErrInvalidMethod)
```
Comment (dadgar) [inline command/agent/operator_endpoint.go]: 404 -> 405

File: command/agent/operator_endpoint.go [hunk 27]
```diff
@@ -208,3 +208,73 @@ func (s *HTTPServer) OperatorServerHealth(resp http.ResponseWriter, req *http.Re
...
+		var reply bool
+		if err := s.agent.RPC("Operator.SchedulerSetConfiguration", &args, &reply); err != nil {
+			return nil, err
+		}
+
+		// Only use the out value if this was a CAS
+		if !args.CAS {
+			return true, nil
+		}
+		return reply, nil
```
Comment (dadgar) [inline command/agent/operator_endpoint.go]: Use SetIndex: https://github.com/hashicorp/nomad/blob/master/command/agent/job_endpoint.go#L396

File: command/agent/operator_endpoint.go [hunk 28]
```diff
@@ -208,3 +208,73 @@ func (s *HTTPServer) OperatorServerHealth(resp http.ResponseWriter, req *http.Re
...
+		out := api.SchedulerConfiguration{
+			PreemptionConfig: api.PreemptionConfig{SystemSchedulerEnabled: reply.SchedulerConfig.PreemptionConfig.SystemSchedulerEnabled},
+			CreateIndex:      reply.CreateIndex,
+			ModifyIndex:      reply.ModifyIndex,
+		}
+
+		resp := api.SchedulerConfigurationResponse{
+			SchedulerConfig: out,
+			CreateIndex:     out.CreateIndex,
+			ModifyIndex:     out.ModifyIndex,
```
Comment (dadgar) [inline command/agent/operator_endpoint.go]: SetMeta: https://github.com/hashicorp/nomad/blob/master/command/agent/job_endpoint.go#L341

File: command/agent/operator_endpoint_test.go [hunk 29]
```diff
@@ -257,3 +258,102 @@ func TestOperator_ServerHealth_Unhealthy(t *testing.T) {
 		})
 	})
 }
+
+func TestOperator_SchedulerGetConfiguration(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (dadgar) [reply command/agent/operator_endpoint_test.go]: Still needs ACL tests

File: nomad/fsm_test.go [hunk 30]
```diff
@@ -2825,3 +2891,49 @@ func TestFSM_Autopilot(t *testing.T) {
...
+	req.CAS = true
+	req.Config.PreemptionConfig = structs.PreemptionConfig{SystemSchedulerEnabled: false}
+	req.Config.ModifyIndex = config.ModifyIndex - 1
+	buf, err = structs.Encode(structs.SchedulerConfigRequestType, req)
+	require.Nil(err)
+
+	resp = fsm.Apply(makeLog(buf))
+	if _, ok := resp.(error); ok {
+		t.Fatalf("bad: %v", resp)
+	}
```
Comment (dadgar) [inline nomad/fsm_test.go]: Repeat the above type asserting to boolean and checking that it is false?

File: nomad/leader.go [hunk 31]
```diff
@@ -42,6 +42,9 @@ const (
 
 var minAutopilotVersion = version.Must(version.NewVersion("0.8.0"))
 
+// Default configuration for scheduler with preemption emabled for system jobs
+var defaultSchedulerConfig = &structs.SchedulerConfiguration{PreemptionConfig: structs.PreemptionConfig{SystemSchedulerEnabled: true}}
```
Comment (dadgar) [inline nomad/leader.go]: Split this into multiple lines

File: nomad/leader.go [hunk 32]
```diff
@@ -42,6 +42,9 @@ const (
 
 var minAutopilotVersion = version.Must(version.NewVersion("0.8.0"))
 
+// Default configuration for scheduler with preemption emabled for system jobs
```
Comment (dadgar) [inline nomad/leader.go]: emabled -> enabled.

File: nomad/leader.go [hunk 33]
```diff
@@ -1230,3 +1236,24 @@ func (s *Server) getOrCreateAutopilotConfig() *structs.AutopilotConfig {
 
 	return config
 }
+
+// getOrCreateSchedulerConfig is used to get the scheduler config, initializing it if necessary
```
Comment (dadgar) [inline nomad/leader.go]: Expand the comment a little and say in a fresh cluster, there won't be a  scheduler config so we need to initialize it

File: nomad/operator_endpoint.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (dadgar) [reply nomad/operator_endpoint.go]: I would still change this. We have talked about the downside of not having flexibility on the endpoints pretty extensively

File: nomad/operator_endpoint.go [hunk 34]
```diff
@@ -284,3 +284,65 @@ func (op *Operator) ServerHealth(args *structs.GenericRequest, reply *autopilot.
...
+	} else if config == nil {
+		return fmt.Errorf("scheduler config not initialized yet")
+	}
+
+	resp := &structs.SchedulerConfigurationResponse{
+		SchedulerConfig: *config,
+		CreateIndex:     config.CreateIndex,
+		ModifyIndex:     config.ModifyIndex,
+	}
+	*reply = *resp
```
Comment (dadgar) [inline nomad/operator_endpoint.go]: Just use the reply, there is no need for the resp variable?

File: nomad/state/state_store.go [hunk 35]
```diff
@@ -3820,6 +3859,84 @@ func (s *StateStore) BootstrapACLTokens(index, resetIndex uint64, token *structs
...
+	s.schedulerSetConfigTxn(idx, tx, config)
+
+	tx.Commit()
+	return nil
+}
+
+// SchedulerCASConfig is used to try updating the scheduler configuration with a
+// given Raft index. If the CAS index specified is not equal to the last observed index
+// for the config, then the call is a noop,
```
Comment (dadgar) [inline nomad/state/state_store.go]: Last comma should be a period

File: nomad/state/state_store.go [hunk 36]
```diff
@@ -3820,6 +3859,84 @@ func (s *StateStore) BootstrapACLTokens(index, resetIndex uint64, token *structs
...
+	tx := s.db.Txn(true)
+	defer tx.Abort()
+
+	s.schedulerSetConfigTxn(idx, tx, config)
+
+	tx.Commit()
+	return nil
+}
+
+// SchedulerCASConfig is used to try updating the scheduler configuration with a
```
Comment (dadgar) [inline nomad/state/state_store.go]: is used to update the scheduler configuration...

File: nomad/state/state_store.go [hunk 37]
```diff
@@ -3820,6 +3859,84 @@ func (s *StateStore) BootstrapACLTokens(index, resetIndex uint64, token *structs
...
+	existing, err := tx.First("scheduler_config", "id")
+	if err != nil {
+		return false, fmt.Errorf("failed scheduler config lookup: %s", err)
+	}
+
+	// If the existing index does not match the provided CAS
+	// index arg, then we shouldn't update anything and can safely
+	// return early here.
+	e, ok := existing.(*structs.SchedulerConfiguration)
+	if !ok || e.ModifyIndex != cidx {
```
Comment (dadgar) [inline nomad/state/state_store.go]: I may be wrong here butt can't `e == nil` and `ok == true`, in which case you can panic dereferencing the modify index.

File: nomad/structs/operator.go [hunk 38]
```diff
@@ -119,3 +119,43 @@ type AutopilotConfig struct {
...
+	ModifyIndex uint64
+}
+
+// SchedulerConfigurationResponse is the response object that wraps SchedulerConfiguration
+type SchedulerConfigurationResponse struct {
+	// SchedulerConfig contains scheduler config options
+	SchedulerConfig SchedulerConfiguration
+
+	// CreateIndex/ModifyIndex store the create/modify indexes of this configuration.
+	CreateIndex uint64
```
Comment (dadgar) [inline nomad/structs/operator.go]: QueryMeta

File: scheduler/preemption.go [hunk 39]
```diff
@@ -0,0 +1,572 @@
...
+				distance := scoreForTaskGroup(resourcesNeeded, alloc.ComparableResources(), maxParallel, currentPreemptionCount)
+				if distance < bestDistance {
+					bestDistance = distance
+					closestAllocIndex = index
+				}
+			}
+			closestAlloc := allocGrp.allocs[closestAllocIndex]
+
+			if preemptedResources == nil {
+				preemptedResources = closestAlloc.ComparableResources()
```
Comment (dadgar) [inline scheduler/preemption.go]: You don't need both preemptedResources and availableResources. Just initialized availableResources to nodeRemainingResources and then add the closest to it as you mark them for preemption

File: scheduler/preemption.go [hunk 40]
```diff
@@ -0,0 +1,572 @@
...
+			break
+		}
+	}
+
+	// Early return if all allocs examined and requirements were not met
+	if !allRequirementsMet {
+		return nil
+	}
+
+	resourcesNeeded = resourceAsk.Comparable()
```
Comment (dadgar) [inline scheduler/preemption.go]: move to be above filteredBestAllocs

File: scheduler/preemption.go [hunk 41]
```diff
@@ -0,0 +1,572 @@
...
+			bestDistance := math.MaxFloat64
+			// Find the alloc with the closest distance
+			for index, alloc := range allocGrp.allocs {
+				currentPreemptionCount := p.getNumPreemptions(alloc)
+				maxParallel := 0
+				tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
+				if tg != nil && tg.Migrate != nil {
+					maxParallel = tg.Migrate.MaxParallel
+				}
+				distance := scoreForTaskGroup(resourcesNeeded, alloc.ComparableResources(), maxParallel, currentPreemptionCount)
```
Comment (dadgar) [inline scheduler/preemption.go]: Try to reduce the number of objects we create and lookups. Can the alloc group have a map of alloc ID to migrate maxparallel and another to the comparable resources.

File: scheduler/preemption.go [hunk 42]
```diff
@@ -0,0 +1,572 @@
...
+	// Early return if there are no current allocs
+	if len(p.currentAllocs) == 0 {
+		return nil
+	}
+
+	deviceToAllocs := make(map[string][]*structs.Allocation)
+	MbitsNeeded := networkResourceAsk.MBits
+	reservedPortsNeeded := networkResourceAsk.ReservedPorts
+
+	// Create a map from each device to allocs
```
Comment (dadgar) [inline scheduler/preemption.go]: It isn't super obvious what the comment blocks are pointing to. I think this one should be above the for loop and the one below should be above the job priority filterign

File: scheduler/preemption.go [hunk 43]
```diff
@@ -0,0 +1,572 @@
...
+		}
+	}
+
+	// If no existing allocations use network resources, return early
+	if len(deviceToAllocs) == 0 {
+		return nil
+	}
+
+	var allocsToPreempt []*structs.Allocation
+
```
Comment (dadgar) [inline scheduler/preemption.go]: Delete blank line

File: scheduler/preemption.go [hunk 44]
```diff
@@ -0,0 +1,572 @@
...
+				}
+			}
+
+			// Look for allocs that are using reserved ports needed
+			for _, port := range reservedPortsNeeded {
+				alloc, ok := usedPortToAlloc[port.Value]
+				if ok {
+					preemptedBandwidth += alloc.Resources.Networks[0].MBits
+					allocsToPreempt = append(allocsToPreempt, alloc)
+				}
```
Comment (dadgar) [inline scheduler/preemption.go]: If !ok can't you continue on the outer for loop since there is a port that can't be acquired (alloc is too high of priority)

File: scheduler/preemption.go [hunk 45]
```diff
@@ -0,0 +1,572 @@
...
+				return p.distanceComparatorForNetwork(allocs, networkResourceAsk, i, j)
+			})
+
+			// Iterate over allocs until end of if requirements have been met
+			for _, alloc := range allocs {
+				preemptedBandwidth += alloc.Resources.Networks[0].MBits
+				allocsToPreempt = append(allocsToPreempt, alloc)
+				if preemptedBandwidth+freeBandwidth >= MbitsNeeded {
+					met = true
+					break
```
Comment (dadgar) [inline scheduler/preemption.go]: Based on my other comment you will have to add:

```
OUTER:
  for device, currentAllocs := range deviceToAllocs {
```

in which case you can remove all these `if met` lines and just do a `break OUTER`

File: scheduler/preemption.go [hunk 46]
```diff
@@ -0,0 +1,572 @@
...
+			if met {
+				break
+			}
+		}
+		// If we met bandwidth needs we don't need to examine the next network device
+		if met {
+			break
+		}
+	}
+	if len(allocsToPreempt) == 0 {
```
Comment (dadgar) [inline scheduler/preemption.go]: This isn't correct. If you go through all devices and the final device you add some allocs to allocsToPreempt but never meet the required mbits this slice won't be empty but you haven't met the resource request

File: scheduler/preemption.go [hunk 47]
```diff
@@ -0,0 +1,572 @@
...
+		// Reset allocsToPreempt since we don't want to preempt across devices for the same task
+		allocsToPreempt = nil
+
+		// Build map from used reserved ports to allocation
+		usedPortToAlloc := make(map[int]*structs.Allocation)
+
+		// First try to satisfy needed reserved ports
+		if len(reservedPortsNeeded) > 0 {
+			for _, alloc := range currentAllocs {
+				for _, tr := range alloc.TaskResources {
```
Comment (dadgar) [inline scheduler/preemption.go]: Shouldn't you be using comparable here. This doesn't use the new AllocatedResources stanza and will break once TaskResources is deprecated

File: scheduler/preemption.go [hunk 48]
```diff
@@ -0,0 +1,572 @@
...
+	mbitsAvailable := n.availableResources.MBits
+	mbitsNeeded := n.resourceNeeded.MBits
+	if mbitsAvailable == 0 || mbitsNeeded == 0 {
+		return false
+	}
+	return mbitsAvailable >= mbitsNeeded
+}
+
+func (n *NetworkPreemptionResource) Distance() float64 {
+	networkCoord := math.MaxFloat64
```
Comment (dadgar) [inline scheduler/preemption.go]: Call networkResourceDistance() to be DRY

File: scheduler/preemption.go [hunk 49]
```diff
@@ -0,0 +1,572 @@
...
+}
+
+// networkResourceDistance returns a distance based only on network megabits
+func networkResourceDistance(resourceUsed *structs.NetworkResource, resourceNeeded *structs.NetworkResource) float64 {
+	networkCoord := math.MaxFloat64
+	if resourceUsed != nil && resourceNeeded != nil {
+		networkCoord = float64(resourceNeeded.MBits-resourceUsed.MBits) / float64(resourceNeeded.MBits)
+	}
+
+	originDist := math.Sqrt(
```
Comment (dadgar) [inline scheduler/preemption.go]: Same issue

File: scheduler/preemption.go [hunk 50]
```diff
@@ -0,0 +1,572 @@
...
+		return distance1 > distance2
+	})
+
+	var preemptedResources *structs.ComparableResources
+	var filteredBestAllocs []*structs.Allocation
+
+	// Do another pass to eliminate allocations that are a superset of other allocations
+	// in the preemption set
+	for _, alloc := range bestAllocs {
+		if preemptedResources == nil {
```
Comment (dadgar) [inline scheduler/preemption.go]: Same comment, you are just accumulating node free resources + preempted resources so just use one object

File: scheduler/preemption.go [hunk 51]
```diff
@@ -0,0 +1,572 @@
...
+func (p *Preemptor) distanceComparatorForNetwork(allocs []*structs.Allocation, networkResourceAsk *structs.NetworkResource, i int, j int) bool {
+	firstAlloc := allocs[i]
+	currentPreemptionCount1 := p.getNumPreemptions(firstAlloc)
+	// Look up configured maxParallel value for these allocation's task groups
+	var maxParallel1, maxParallel2 int
+	tg1 := allocs[i].Job.LookupTaskGroup(firstAlloc.TaskGroup)
+	if tg1 != nil && tg1.Migrate != nil {
+		maxParallel1 = tg1.Migrate.MaxParallel
+	}
+	// Dereference network usage on first alloc if its there
```
Comment (dadgar) [inline scheduler/preemption.go]: Blank lines above comments

File: scheduler/preemption.go [hunk 52]
```diff
@@ -0,0 +1,572 @@
...
+		c := countMap[alloc.TaskGroup]
+		countMap[alloc.TaskGroup] = c + 1
+		p.currentPreemptions[id] = countMap
+	}
+}
+
+// getNumPreemptions counts the number of other allocations being preempted that match the job and task group of
+// the alloc under consideration. This is used as a scoring factor to minimize too many allocs of the same job being preempted at once
+func (p *Preemptor) getNumPreemptions(alloc *structs.Allocation) int {
+	numCurrentPreemptionsForJob := 0
```
[the remark this answers is not in the document]
Comment (preetapan) [reply scheduler/preemption.go]: TIL

File: scheduler/preemption.go (hunk 44 above)
Comment (preetapan) [reply scheduler/preemption.go]: I'll add a test case for this, good catch

File: nomad/plan_apply.go [hunk 53]
```diff
@@ -318,6 +359,25 @@ func evaluatePlanPlacements(pool *EvaluatePool, snap *state.StateSnapshot, plan
 		if nodeAlloc := plan.NodeAllocation[nodeID]; len(nodeAlloc) > 0 {
 			result.NodeAllocation[nodeID] = nodeAlloc
 		}
+
+		if nodePreemptions := plan.NodePreemptions[nodeID]; nodePreemptions != nil {
+			var filteredNodePreemptions []*structs.Allocation
+			// Do a pass over preempted allocs in the plan to check
+			// whether the alloc is already in a terminal state
+			for _, preemptedAlloc := range nodePreemptions {
```
[the remark this answers is not in the document]
Comment (preetapan) [reply nomad/plan_apply.go]: added unit test in plan_apply_test.go

File: scheduler/preemption.go [hunk 54]
```diff
@@ -395,10 +428,23 @@ func (p *Preemptor) PreemptForNetwork(networkResourceAsk *structs.NetworkResourc
 			Networks: []*structs.NetworkResource{networkResourceAsk},
 		},
 	}
-	filteredBestAllocs := filterSuperset(allocsToPreempt, nodeRemainingResources, resourcesNeeded, preemptionResourceFactory)
+	filteredBestAllocs := p.filterSuperset(allocsToPreempt, nodeRemainingResources, resourcesNeeded, preemptionResourceFactory)
 	return filteredBestAllocs
 }
 
+func usedReservedPorts(net *structs.NetworkResource, portMap map[int]interface{}) bool {
```
Comment (dadgar) [inline scheduler/preemption.go]: Deadcode?

File: scheduler/preemption.go [hunk 55]
```diff
@@ -0,0 +1,617 @@
...
+					bestDistance = distance
+					closestAllocIndex = index
+				}
+			}
+			closestAlloc := allocGrp.allocs[closestAllocIndex]
+			closestResources := p.allocDetails[closestAlloc.ID].resources
+			availableResources.Add(closestResources)
+
+			// This step needs the original resources asked for as the second arg, can't use the running total
+			allRequirementsMet, _ = availableResources.Superset(resourceAsk.Comparable())
```
Comment (dadgar) [inline scheduler/preemption.go]: compute `resourceAsk.Comparable()` once at the top of this method?

File: scheduler/preemption.go [hunk 56]
```diff
@@ -0,0 +1,617 @@
...
+	if len(p.currentAllocs) == 0 {
+		return nil
+	}
+
+	deviceToAllocs := make(map[string][]*structs.Allocation)
+	MbitsNeeded := networkResourceAsk.MBits
+	reservedPortsNeeded := networkResourceAsk.ReservedPorts
+
+	// Build map of reserved ports needed for fast access
+	reservedPorts := make(map[int]interface{})
```
Comment (dadgar) [inline scheduler/preemption.go]: Use struct{} instead of interface{}. It is a smaller object and more canonical for set implementation.

File: scheduler/preemption.go [hunk 57]
```diff
@@ -0,0 +1,617 @@
...
+	for _, port := range reservedPortsNeeded {
+		reservedPorts[port.Value] = struct{}{}
+	}
+
+	filteredReservedPorts := make(map[string]map[int]struct{})
```
Comment (dadgar) [inline scheduler/preemption.go]: Comment on what this map is for

File: scheduler/preemption.go [hunk 58]
```diff
@@ -0,0 +1,617 @@
...
+
+		// Track how much existing free bandwidth we have before preemption
+		freeBandwidth = totalBandwidth - netIdx.UsedBandwidth[device]
+
+		preemptedBandwidth := 0
+
+		// Reset allocsToPreempt since we don't want to preempt across devices for the same task
+		allocsToPreempt = nil
+
+		usedPortToAlloc := make(map[int]*structs.Allocation)
```
Comment (dadgar) [inline scheduler/preemption.go]: Comment

**Comment by dadgar** [comment]
LGTM after last comments are addressed
